From 5dcc08a9960329cdadba297a3c916edfb8ccc3ad Mon Sep 17 00:00:00 2001
From: dfffffff <dflamand@gmail.com>
Date: Wed, 27 Jul 2016 13:03:41 -0400
Subject: core: register copy propagation issue:

This patch is based on the work of 
David Flamand https://gitlab.com/dfffffff/gcc6809

applied patch from https://gcc.gnu.org/viewcvs/gcc/trunk/gcc/regcprop.c?view=patch&r1=205117&r2=205116&pathrev=205117

NOTE: The PR number of 54300 is wrong; but, it is still the best thing to search on.
	PR rtl-optimization/54300
	* regcprop.c (copyprop_hardreg_forward_1): Ensure any unused
	outputs in a single-set are killed from the value chains.

---
 gcc/regrename.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/gcc/regrename.c b/gcc/regrename.c
index 0d21ae2e8b8..02c962ee1b6 100644
--- a/gcc/regrename.c
+++ b/gcc/regrename.c
@@ -1579,6 +1579,7 @@ copyprop_hardreg_forward_1 (basic_block bb, struct value_data *vd)
 {
   bool changed = false;
   rtx insn;
+  rtx link;
 
   for (insn = BB_HEAD (bb); ; insn = NEXT_INSN (insn))
     {
@@ -1638,6 +1639,23 @@ copyprop_hardreg_forward_1 (basic_block bb, struct value_data *vd)
 	if (recog_op_alt[i][alt].earlyclobber)
 	  kill_value (recog_data.operand[i], vd);
 
+      /* If we have dead sets in the insn, then we need to note these as we
+         would clobbers.  */
+      for (link = REG_NOTES (insn); link; link = XEXP (link, 1))
+        {
+          if (REG_NOTE_KIND (link) == REG_UNUSED)
+            {
+              kill_value (XEXP (link, 0), vd);
+              /* Furthermore, if the insn looked like a single-set,
+                 but the dead store kills the source value of that
+                 set, then we can no-longer use the plain move
+                 special case below.  */
+              if (set
+                  && reg_overlap_mentioned_p (XEXP (link, 0), SET_SRC (set)))
+                set = NULL;
+            }
+        }
+
       /* Special-case plain move instructions, since we may well
 	 be able to do the move from a different register class.  */
       if (set && REG_P (SET_SRC (set)))
-- 
2.19.1.windows.1

